website/blog/en/2021/clickhouse-v21.12-released.md

```diff
@@ -0,0 +1,60 @@
+
+
+## Table constraints
+
+When creating tables, you can now (optionally) specify constraints. A constraint tells ClickHouse that a column has a specific relationship to another column in the same table. For example, a string column might store the prefix of another column. Then, when a select query is trying to calculate the prefix on the original column, ClickHouse will rewrite the query to use the prefix column.
+
+For now, constraints are implemented as assumptions, that is, ClickHouse does not validate that they are correct and will not reject new data that violates them.
+
+This feature is disabled by default. To turn it on, enable `optimize_using_constraints`, `optimize_substitute_columns` and/or `optimize_append_index`.
+
+**How does this help you?**
+
+Especially in large ClickHouse deployments with many complex tables it can be hard for users to always be up to date on the best way to query a given dataset. Constraints can help optimize queries without having to change the query structure itself. They can also make it easier to make changes to tables. For example, let’s say you have a table containing web requests and it includes a URL column that contains the full URL of each request. Many times, users will want to know the top level domain (.com, .co.uk, etc.), something ClickHouse provides the topLevelDomain function to calculate. If you discover that many people are using this function you might decide to create a new materialized column that pre-calculates the top level domain for each record. Rather than tell all your users to change their queries you can use a table constraint to tell ClickHouse that each time a user tries to call the topLevelDomain function the request should be rewritten to use the new materialized column.
+
+
+## Read large remote files in chunks
+
+When reading large files in Parquet, ORC, and Arrow format using the s3, url, and hdfs table functions, ClickHouse will now automatically choose whether to read the entire file at once or read parts of it incrementally. This is now enabled by default and the setting `remote_read_min_bytes_for_seek` controls when to switch from reading it all to reading in chunks. The default is 1MiB.
+
+**How does this help our ClickHouse Users?**
+
+In previous versions, when reading files from remote locations with the s3, url, and hdfs table functions, ClickHouse would always read the entire file into memory. This works well when the files are small but will cause excessive memory usage or not work at all when the files are large. With this change, ClickHouse will read large files in chunks to keep memory usage in check and is now able to read even very large files. 
+
+ClickHouse Keeper is 100% feature complete. More updates to come in the coming weeks around where and how you can test and provide feedback for us!
```

Comment:
looks weird as this doesn't have it's own `<h2>` and is completely unrelated to the previous header